File: m/arrays.m
%------------------------------------------------------------------------------%
% ARRAYS

% 3D vectors
w1   = repmat( zero, [ nm 3 ] ); % temporary storage
w2   = w1;                       % temporary storage
x    = w1;                       % node locations
v    = w1;                       % **velocity
u    = w1;                       % **displacement

% 3D scalars
s1   = repmat( zero, nm );       % temporary storage
s2   = s1;                       % temporary storage
mr   = s1;                       % mass ratio
lam  = s1;                       % Lame parameter
mu   = s1;                       % Lame parameter
y    = s1;                       % hourglass constant

if ifn
  nf = nm;
  nf(ifn) = 1;
else
  nf = [ 0 0 0 ];
end

% Fault vetors
t1   = repmat( zero, [ nf 3 ] ); % temporary storage
t2   = t1;                       % temporary storage
t3   = t1;                       % temporary storage
nrm  = t1;                       % fault normal vector
t0   = t1;                       % initial traction

% Fault scalars
f1   = repmat( zero, nf );       % temporary storage
f2   = f1;                       % temporary storage
tn   = f1;                       % temporary storage
ts   = f1;                       % temporary storage
mus  = f1;                       % coef of static friction
mud  = f1;                       % coef of dynamic friction
dc   = f1;                       % slip weakening distance
co   = f1;                       % cohesion
area = f1;                       % fault element area
r    = f1;                       % radius to hypocenter
vs   = f1;                       % **slip velocity
us   = f1;                       % **slip
trup = f1;                       % **rupture time

i1 = [ 0 0 0 ];
i2 = [ 0 0 0 ];
i = bc1 == 1; i1(i) = npml;
i = bc2 == 1; i2(i) = npml;
nj1 = [ nm 3 ]; nj1(1) = i1(1);
nj2 = [ nm 3 ]; nj2(1) = i2(1);
nk1 = [ nm 3 ]; nk1(2) = i1(2);
nk2 = [ nm 3 ]; nk2(2) = i2(2);
nl1 = [ nm 3 ]; nl1(3) = i1(3);
nl2 = [ nm 3 ]; nl2(3) = i2(3);

% PML state
p1  = repmat( zero, nj1 );       % **j1 pml momentum
p2  = repmat( zero, nk1 );       % **k1 pml momentum
p3  = repmat( zero, nl1 );       % **l1 pml momentum
p4  = repmat( zero, nj2 );       % **j2 pml momentum
p5  = repmat( zero, nk2 );       % **k2 pml momentum
p6  = repmat( zero, nl2 );       % **l2 pml momentum
g1  = repmat( zero, nj1 );       % **j1 pml gradient
g2  = repmat( zero, nk1 );       % **k1 pml gradient
g3  = repmat( zero, nl1 );       % **l1 pml gradient
g4  = repmat( zero, nj2 );       % **j2 pml gradient
g5  = repmat( zero, nk2 );       % **k2 pml gradient
g6  = repmat( zero, nl2 );       % **l2 pml gradient

% Initial values
t = 0.;
amax  = 0.;
vmax  = 0.;
umax  = 0.;
wmax  = 0.;
vsmax = 0.;
usmax = 0.;
tnmax = 0.;
tsmax = 0.;

File: f/arrays.f90
! Allocate arrays
module m_arrays
implicit none
contains

subroutine arrays
use m_globals
integer :: i1(3), i2(3), j, k, l, j1, k1, l1, j2, k2, l2

i2 = nm
j = i2(1)
k = i2(2)
l = i2(3)

! 3d vectors
allocate(       &
  v(j,k,l,3),   &
  u(j,k,l,3),   &
! z1(j,k,l,3),  &
! z2(j,k,l,3),  &
  w1(j,k,l,3),  &
  w2(j,k,l,3)   )

! 3d scalars
allocate(       &
  mr(j,k,l),    &
  lam(j,k,l),   &
  mu(j,k,l),    &
  gam(j,k,l),   &
! qp(j,k,l),    &
! qs(j,k,l),    &
  y(j,k,l),     &
  pv(j,k,l),    &
  s1(j,k,l),    &
  s2(j,k,l)     )

i1 = 0
i2 = 0
where ( bc1 == 1 ) i1 = npml
where ( bc2 == 1 ) i2 = npml
j1 = i1(1); j2 = i2(1)
k1 = i1(2); k2 = i2(2)
l1 = i1(3); l2 = i2(3)

! PML state
allocate(       &
  p1(j1,k,l,3), &
  g1(j1,k,l,3), &
  p2(j,k1,l,3), &
  g2(j,k1,l,3), &
  p3(j,k,l1,3), &
  g3(j,k,l1,3), &
  p4(j2,k,l,3), &
  g4(j2,k,l,3), &
  p5(j,k2,l,3), &
  g5(j,k2,l,3), &
  p6(j,k,l2,3), &
  g6(j,k,l2,3)  )

! PML damping
allocate( dn1(npml), dn2(npml), dc1(npml), dc2(npml) )

if ( ifn /= 0 ) then
  i1 = nm
  i1(ifn) = 1
else
  i1 = 0
end if
j = i1(1)
k = i1(2)
l = i1(3)

! Fault vectors
allocate(        &
  nhat(j,k,l,3), &
  t0(j,k,l,3),   &
  t1(j,k,l,3),   &
  t2(j,k,l,3),   &
  t3(j,k,l,3)    )

! Fault scalars
allocate(       &
  mus(j,k,l),   &
  mud(j,k,l),   &
  dc(j,k,l),    &
  co(j,k,l),    &
  area(j,k,l),  &
  rhypo(j,k,l), &
  muf(j,k,l),   &
  sl(j,k,l),    &
  psv(j,k,l),   &
  trup(j,k,l),  &
  tarr(j,k,l),  &
  tn(j,k,l),    &
  ts(j,k,l),    &
  f1(j,k,l),    &
  f2(j,k,l)     )

end subroutine

end module

